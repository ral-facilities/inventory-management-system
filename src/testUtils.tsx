import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render } from '@testing-library/react';
import { enGB } from 'date-fns/locale/en-GB';
import React from 'react';
import {
  RouterProvider,
  createBrowserRouter,
  createMemoryRouter,
} from 'react-router-dom';
import { paths } from './App';
import {
  CatalogueCategory,
  CatalogueCategoryProperty,
  CatalogueItem,
  Item,
  Manufacturer,
} from './api/api.types';
import CatalogueCategoriesJSON from './mocks/CatalogueCategories.json';
import CatalogueItemsJSON from './mocks/CatalogueItems.json';
import ItemsJSON from './mocks/Items.json';
import ManufacturersJSON from './mocks/Manufacturers.json';

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

function constructRouterProvider(
  ui: React.ReactNode,
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const Root: React.FunctionComponent = () => {
    return (
      <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
        <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
      </LocalizationProvider>
    );
  };

  const router =
    initialEntry !== undefined && urlPathKey !== undefined
      ? createMemoryRouter([{ path: paths[urlPathKey], Component: Root }], {
          initialEntries: [initialEntry],
        })
      : createBrowserRouter([{ path: '*', Component: Root }]);

  return { router: router, provider: <RouterProvider router={router} /> };
}

function constructRouterProviderWrapper(
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const wrapper = ({
    children,
  }: {
    children: React.ReactNode;
  }): JSX.Element => {
    return constructRouterProvider(
      children,
      queryClient,
      urlPathKey,
      initialEntry
    ).provider;
  };
  return wrapper;
}

export function renderComponentWithRouterProvider(
  ui: React.ReactElement,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string,
  {
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const { router, provider } = constructRouterProvider(
    ui,
    queryClient,
    urlPathKey,
    initialEntry
  );
  return {
    queryClient,
    router,
    ...render(ui, {
      wrapper: () => provider,
      ...renderOptions,
    }),
  };
}

export const hooksWrapperWithProviders = (props?: {
  queryClient?: QueryClient;
  urlPathKey?: keyof typeof paths;
  initialEntry?: string;
}) => {
  const testQueryClient = props?.queryClient ?? createTestQueryClient();
  return constructRouterProviderWrapper(
    testQueryClient,
    props?.urlPathKey,
    props?.initialEntry
  );
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryProperty[] => {
  const filteredCategories = CatalogueCategoriesJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].properties ?? [];
  return properties as CatalogueCategoryProperty[];
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoriesJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemsJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemsJSON.filter(
    (catalogueItem) => catalogueItem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturersJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemById = (itemId: string): Item | undefined => {
  return ItemsJSON.find((item) => item.id === itemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};

export const CREATED_MODIFIED_TIME_VALUES = {
  created_time: '2024-01-01T12:00:00.000+00:00',
  modified_time: '2024-01-02T13:10:10.000+00:00',
};
