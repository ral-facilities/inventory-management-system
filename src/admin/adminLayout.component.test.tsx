import { screen, waitFor } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { urlPathKeyType } from '../paths';
import { renderComponentWithRouterProvider } from '../testUtils';
import AdminLayout, { AdminErrorComponent } from './adminLayout.component';

const mockedUseNavigate = vi.fn();

vi.mock('react-router-dom', async () => ({
  ...(await vi.importActual('react-router-dom')),
  useNavigate: () => mockedUseNavigate,
}));

describe('Admin Layout', () => {
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
  });
  const createView = (path: string, urlPathKey: urlPathKeyType) => {
    return renderComponentWithRouterProvider(<AdminLayout />, urlPathKey, path);
  };

  it('renders admin layout page correctly', async () => {
    const view = createView('/admin-ims', 'admin');

    await waitFor(() => {
      expect(
        screen.getByRole('button', {
          name: 'navigate to admin home',
        })
      ).toBeInTheDocument();
    });

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders units breadcrumbs correctly', async () => {
    const view = createView('/admin-ims/units', 'adminUnits');

    await waitFor(() => {
      expect(screen.getByText('Units')).toBeInTheDocument();
    });

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('renders usage statuses breadcrumbs correctly', async () => {
    const view = createView('/admin-ims/usage-statuses', 'adminUsageStatuses');

    await waitFor(() => {
      expect(screen.getByText('Usage statuses')).toBeInTheDocument();
    });

    expect(view.asFragment()).toMatchSnapshot();
  });

  it('calls useNavigate when the home button is clicked', async () => {
    createView('/admin-ims/usage-statuses', 'adminUsageStatuses');

    await waitFor(() => {
      expect(screen.getByText('Usage statuses')).toBeInTheDocument();
    });

    const homeButton = screen.getByRole('button', {
      name: 'navigate to admin home',
    });

    await user.click(homeButton);

    expect(mockedUseNavigate).toHaveBeenCalledTimes(1);
    expect(mockedUseNavigate).toHaveBeenCalledWith('/admin-ims');
  });
});

describe('Admin Error Component', () => {
  const createView = () => {
    return renderComponentWithRouterProvider(<AdminErrorComponent />);
  };

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders Admin error page correctly', async () => {
    const view = createView();

    expect(view.asFragment()).toMatchSnapshot();
  });
});
