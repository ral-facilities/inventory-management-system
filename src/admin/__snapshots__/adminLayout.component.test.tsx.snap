// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Admin Error Component > renders Admin error page correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-7gabc3"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-11gsi37-MuiTypography-root"
    >
      Invalid Admin Route
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
    >
      The admin route you are trying to access doesn't exist. Please click the Home button to navigate back to the Admin Home page.
    </p>
  </div>
</DocumentFragment>
`;

exports[`Admin Layout > renders admin layout page correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-v2kfba"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1r501zp-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <span
          aria-label="Admin Home"
          class=""
          data-mui-internal-clone-element="true"
        >
          <button
            aria-label="navigate to admin home"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </span>
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1mc59t7-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          />
        </nav>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Admin Layout > renders units breadcrumbs correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-v2kfba"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1r501zp-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <span
          aria-label="Admin Home"
          class=""
          data-mui-internal-clone-element="true"
        >
          <button
            aria-label="navigate to admin home"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </span>
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1mc59t7-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <div
                class="MuiBox-root css-8zy5qf"
              />
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <p
                aria-label="Units"
                class="MuiTypography-root MuiTypography-body1 css-1qs6ojv-MuiTypography-root"
                data-mui-internal-clone-element="true"
                role="tooltip"
              >
                Units
              </p>
            </li>
          </ol>
        </nav>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Admin Layout > renders usage statuses breadcrumbs correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-v2kfba"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-1r501zp-MuiGrid-root"
    >
      <div
        class="MuiBox-root css-70qvj9"
      >
        <span
          aria-label="Admin Home"
          class=""
          data-mui-internal-clone-element="true"
        >
          <button
            aria-label="navigate to admin home"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </span>
        <nav
          aria-label="breadcrumb"
          class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-1mc59t7-MuiTypography-root-MuiBreadcrumbs-root"
        >
          <ol
            class="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <div
                class="MuiBox-root css-8zy5qf"
              />
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <p
                aria-label="Usage statuses"
                class="MuiTypography-root MuiTypography-body1 css-1qs6ojv-MuiTypography-root"
                data-mui-internal-clone-element="true"
                role="tooltip"
              >
                Usage statuses
              </p>
            </li>
          </ol>
        </nav>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
